planner/core/tiflash_selection_late_materialization.go

```diff
@@ -0,0 +1,257 @@
+		selectivity, _, err := physicalTableScan.tblColHists.Selectivity(sctx, group, nil)
+		if err != nil {
+			logutil.BgLogger().Warn("calculate selectivity failed, do not push down the conditions group", zap.Error(err))
+			continue
+		}
+		if selectivity <= selectivityThreshold {
+			exprGroups = append(exprGroups, expressionGroup{exprs: group, selectivity: selectivity})
+		}
+	}
+
+	// Sort exprGroups by selectivity in ascending order
+	sort.SliceStable(exprGroups, func(i, j int) bool {
+		return exprGroups[i].selectivity < exprGroups[j].selectivity || (exprGroups[i].selectivity == exprGroups[j].selectivity && len(exprGroups[i].exprs) < len(exprGroups[j].exprs))
+	})
+
+	return exprGroups
+}
+
+// withHeavyCostFunction is used to check if the condition contain heavy cost functions.
+// @param: cond: condition of PhysicalSelection to be checked
+// @note: heavy cost functions are functions that may cause a lot of memory allocation or disk IO.
+func withHeavyCostFunction(cond expression.Expression) bool {
+	if binop, ok := cond.(*expression.ScalarFunction); ok {
+		switch binop.FuncName.L {
+		// JSON functions
+		case ast.JSONType, ast.JSONExtract, ast.JSONUnquote, ast.JSONArray, ast.JSONObject, ast.JSONMerge, ast.JSONSet, ast.JSONInsert, ast.JSONReplace, ast.JSONRemove:
+			return true
+		case ast.JSONMemberOf, ast.JSONContainsPath, ast.JSONValid, ast.JSONArrayAppend, ast.JSONArrayInsert, ast.JSONMergePatch, ast.JSONMergePreserve, ast.JSONPretty:
+			return true
+		case ast.JSONStorageFree, ast.JSONStorageSize, ast.JSONDepth, ast.JSONKeys, ast.JSONLength, ast.JSONContains, ast.JSONSearch, ast.JSONOverlaps, ast.JSONQuote:
+			return true
+		// some time functions
+		case ast.AddDate, ast.AddTime, ast.ConvertTz, ast.DateLiteral, ast.DateAdd, ast.DateFormat, ast.FromUnixTime, ast.GetFormat, ast.UTCTimestamp:
+			return true
+		case ast.DateSub, ast.DateDiff, ast.DayOfYear, ast.Extract, ast.FromDays, ast.TimestampLiteral, ast.TimestampAdd, ast.UnixTimestamp:
+			return true
+		case ast.LocalTimestamp, ast.MakeDate, ast.MakeTime, ast.MonthName, ast.PeriodAdd, ast.PeriodDiff, ast.Quarter, ast.SecToTime, ast.ToSeconds:
+			return true
+		case ast.StrToDate, ast.SubDate, ast.SubTime, ast.TimeLiteral, ast.TimeFormat, ast.TimeToSec, ast.TimeDiff, ast.TimestampDiff:
+			return true
+			// TODO: add more heavy cost functions
+		}
+	}
+	return false
+}
+
+// removeSpecificExprsFromSelection is used to remove the conditions that needed to be pushed down.
+// @param: physicalSelection: the PhysicalSelection to be modified
+// @param: exprs: the conditions to be removed
+func removeSpecificExprsFromSelection(physicalSelection *PhysicalSelection, exprs []expression.Expression) {
+	newConds := make([]expression.Expression, 0, len(physicalSelection.Conditions))
+	for _, cond := range physicalSelection.Conditions {
+		if !expression.Contains(exprs, cond) {
+			newConds = append(newConds, cond)
+		}
+	}
+	physicalSelection.Conditions = newConds
```

Comment:
We can do reverse traverse to avoid creating new slice:
```
conds := sel.Condition
for i := len(conds) - 1; i >= 0; i-- {
	if Contains(exprs, cond) {
		conds = append(conds[:i], conds[i+1:]...)
	}
}
sel.Condition = conds
```